File: src/main/java/com/github/ericytsang/research2016/propositionallogic/Proposition.java
package com.github.ericytsang.research2016.propositionallogic;

import com.github.ericytsang.lib.collections.AbstractIterator;
import com.github.ericytsang.lib.collections.Bounds;
import com.github.ericytsang.lib.collections.BranchAndBound;
import com.github.ericytsang.lib.collections.CollectionWrapper;
import com.github.ericytsang.lib.collections.ContainerUtils;
import com.github.ericytsang.lib.collections.Function;
import com.github.ericytsang.lib.collections.IteratorToSetAdapter;
import com.github.ericytsang.lib.collections.NodeMetadata;

import java.io.Serializable;
import java.util.*;

/**
 * Created by surpl on 7/26/2016.
 */
public abstract class Proposition implements Serializable
{
    public static final Companion Companion = new Companion();

    public static class Companion
    {
        private Companion()
        {}
    }

    public static final Proposition TAUTOLOGY = new Operand("1")
    {
        @Override
        boolean truthValue(State state)
        {
            return true;
        }
    };

    public static final Proposition CONTRADICTION = new Operand("0")
    {
        @Override
        boolean truthValue(State state)
        {
            return false;
        }
    };

    public static Proposition makeDnf(final State state)
    {
        return CollectionWrapper.wrap(state.keySet())
            .map(new Function<Variable,Proposition>()
            {
                @Override
                public Proposition invoke(Variable params)
                {
                    if (state.get(params))
                    {
                        return params;
                    }
                    else
                    {
                        return params.not();
                    }
                }
            })
            .let(new Function<CollectionWrapper<List<Proposition>,Proposition>,Proposition>()
            {
                @Override
                public Proposition invoke(CollectionWrapper<List<Proposition>,Proposition> params)
                {
                    Proposition result = And.make(params.wrapee);
                    return result != null ? result : Proposition.TAUTOLOGY;
                }
            });
    }

    public static Proposition makeDnf(Iterable<State> states)
    {
        CollectionWrapper<LinkedList<State>,State> collectionWrapper = CollectionWrapper.wrap(new LinkedList<State>());
        for (State state : states)
        {
            collectionWrapper.add(state);
        }
        return collectionWrapper
            .map(new Function<State,Proposition>()
            {
                @Override
                public Proposition invoke(State params)
                {
                    return Proposition.makeDnf(params);
                }
            })
            .let(new Function<CollectionWrapper<List<Proposition>,Proposition>,Proposition>()
            {
                @Override
                public Proposition invoke(CollectionWrapper<List<Proposition>,Proposition> params)
                {
                    Proposition result = Or.make(params.wrapee);
                    return result != null ? result : Proposition.CONTRADICTION;
                }
            });
    }

    public abstract List<Proposition> getChildren();

    public Set<Variable> getVariables()
    {
        if(this instanceof Variable)
        {
            Set<Variable> set = new HashSet<Variable>(1);
            set.add((Variable) this);
            return set;
        }
        else
        {
            return CollectionWrapper.wrap(getChildren())
                .flatMap(new Function<Proposition,Collection<Variable>>()
                {
                    @Override
                    public Collection<Variable> invoke(Proposition params)
                    {
                        return params.getVariables();
                    }
                })
                .let(new Function<CollectionWrapper<List<Variable>,Variable>,Set<Variable>>()
                {
                    @Override
                    public Set<Variable> invoke(CollectionWrapper<List<Variable>,Variable> params)
                    {
                        return new HashSet<Variable>(params);
                    }
                });
        }
    }

    private Set<State> models = null;

    public Set<State> getModels()
    {
        // if a previous version of models was calculated and stored, return
        // that. calculate otherwise.
        if (models != null)
        {
            return models;
        }

        final BranchAndBound<State> findModel = new BranchAndBound<State>()
        {
            @Override
            protected Set<State> getChildren(final State state)
            {
                final Variable nextVariable = CollectionWrapper.wrap(getVariables())
                    .filter(new Function<Variable,Boolean>()
                    {
                        @Override
                        public Boolean invoke(Variable params)
                        {
                            return !state.containsKey(params);
                        }
                    })
                    .firstOrNull();
                if (nextVariable != null)
                {
                    State node1 = new Function<Void,State>()
                    {
                        @Override
                        public State invoke(Void params)
                        {
                            Map<Variable,Boolean> map = new LinkedHashMap<Variable,Boolean>();
                            map.putAll(state);
                            map.put(nextVariable,true);
                            return State.fromVariableMap(map);
                        }
                    }.invoke(null);
                    State node2 = new Function<Void,State>()
                    {
                        @Override
                        public State invoke(Void params)
                        {
                            Map<Variable,Boolean> map = new LinkedHashMap<Variable,Boolean>();
                            map.putAll(state);
                            map.put(nextVariable,false);
                            return State.fromVariableMap(map);
                        }
                    }.invoke(null);
                    return CollectionWrapper.wrap(ContainerUtils.listOf(node1,node2))
                        .filter(new Function<State,Boolean>()
                        {
                            @Override
                            public Boolean invoke(State params)
                            {
                                return truthiness(params) != 0.0;
                            }
                        })
                        .toSet().wrapee;
                }
                else
                {
                    return Collections.emptySet();
                }
            }

            @Override
            protected Bounds getBounds(State state)
            {
                return new Bounds(1.0,Math.floor(truthiness(state)));
            }

            @Override
            protected Boolean isSolution(State state)
            {
                return truthiness(state) == 1.0 && state.keySet().containsAll(getVariables());
            }
        };

        Iterator<State> iterator = new AbstractIterator<State>()
        {
            private Map<State,NodeMetadata> unbranchedNodes = new LinkedHashMap<State,NodeMetadata>();

            {
                unbranchedNodes.put(State.fromStringMap(Collections.<String,Boolean>emptyMap()),BranchAndBound.ROOT_NODE_METADATA);
            }

            @Override
            protected void computeNext()
            {
                State next = findModel.runAlgorithm(unbranchedNodes);
                if (next == null)
                {
                    done();
                }
                else
                {
                    setNext(next);
                }
            }

            @Override
            protected void doRemove(int indexToRemove,State toRemove)
            {
                throw new UnsupportedOperationException("objects are generated. collection is immutable");
            }
        };

        return new IteratorToSetAdapter<State>(iterator);

    }

    /**
     * evaluates if {@code this} is a subset of {@code that}.
     * @param that the {@code Proposition} instance to test against.
     * @return true if {@code this} is a subset of {@code that}.
     */
    public boolean isSubsetOf(Proposition that)
    {
        return this.not().and(this).isContradiction();
    }

    public boolean isSatisfiedBy(Proposition that)
    {
        if (!isSubsetOf(that))
        {
            return false;
        }

        for (State model : getModels())
        {
            if (!Proposition.makeDnf(model).and(that).isSatisfiable())
            {
                return false;
            }
        }
        return true;
    }

    public boolean isSatisfiable()
    {
        return !getModels().isEmpty();
    }

    public boolean isTautology()
    {
        return !not().isSatisfiable();
    }

    public boolean isContradiction()
    {
        return !isSatisfiable();
    }

    public boolean evaluate(final State state)
    {
        if (this instanceof Operand)
        {
            return ((Operand) this).truthValue(state);
        }
        else if (this instanceof Operator)
        {
            return CollectionWrapper.wrap(getChildren())
                .map(new Function<Proposition,Boolean>()
                {
                    @Override
                    public Boolean invoke(Proposition params)
                    {
                        return params.evaluate(state);
                    }
                })
                .let(new Function<CollectionWrapper<List<Boolean>,Boolean>,Boolean>()
                {
                    @Override
                    public Boolean invoke(CollectionWrapper<List<Boolean>,Boolean> params)
                    {
                        return ((Operator) Proposition.this).operateBooleans(params.wrapee);
                    }
                });
        }
        else
        {
            throw new IllegalArgumentException("unhandled type: "+this);
        }
    }

    public double truthiness(final State state)
    {
        if (this instanceof Variable)
        {
            if (state.keySet().contains((Variable) this))
            {
                if (((Variable) this).truthValue(state))
                {
                    return 1.0;
                }
                else
                {
                    return 0.0;
                }
            }
            else
            {
                return 0.5;
            }
        }
        else if (this instanceof Operand)
        {
            if (this == Proposition.TAUTOLOGY)
            {
                return 1.0;
            }
            else if(this == Proposition.CONTRADICTION)
            {
                return 0.0;
            }
            else
            {
                throw new RuntimeException("unhandled type: "+this);
            }
        }
        else if (this instanceof Operator)
        {
            return CollectionWrapper.wrap(getChildren())
                .map(new Function<Proposition,Double>()
                {
                    @Override
                    public Double invoke(Proposition params)
                    {
                        return params.truthiness(state);
                    }
                })
                .let(new Function<CollectionWrapper<List<Double>,Double>,Double>()
                {
                    @Override
                    public Double invoke(CollectionWrapper<List<Double>,Double> params)
                    {
                        return ((Operator) Proposition.this).operateDoubles(params.wrapee);
                    }
                });
        }
        else
        {
            throw new RuntimeException("unhandled type: "+this);
        }
    }

    public Proposition not()
    {
        return new Not(this);
    }

    public Proposition and(Proposition that)
    {
        ArrayList<Proposition> list = new ArrayList<Proposition>(2);
        list.add(this);
        list.add(that);
        return And.make(list);
    }

    public Proposition or(Proposition that)
    {
        ArrayList<Proposition> list = new ArrayList<Proposition>(2);
        list.add(this);
        list.add(that);
        return Or.make(list);
    }

    public Proposition iff(Proposition that)
    {
        ArrayList<Proposition> list = new ArrayList<Proposition>(2);
        list.add(this);
        list.add(that);
        return Iff.make(list);
    }

    public Proposition oif(Proposition that)
    {
        return new Oif(this,that);
    }

    public Proposition then(Proposition that)
    {
        return new Oif(this,that);
    }

    public Proposition xor(Proposition that)
    {
        return new Xor(this,that);
    }

    public Proposition nand(Proposition that)
    {
        return new Nand(this,that);
    }

    @Override
    public abstract String toString();

    @Override
    public int hashCode()
    {
        return toString().hashCode();
    }

    @Override
    public boolean equals(Object other)
    {
        return other instanceof Proposition && other.toString().equals(toString());
    }
}

File: src/main/kotlin/com/github/ericytsang/research2016/propositionallogic/Proposition.kt
package com.github.ericytsang.research2016.propositionallogic

import com.github.ericytsang.lib.collections.Bounds
import com.github.ericytsang.lib.collections.getRandom
import com.github.ericytsang.lib.collections.IteratorToSetAdapter
import com.github.ericytsang.lib.collections.branchAndBound
import com.github.ericytsang.lib.collections.rootNodeMetaData
import com.github.ericytsang.lib.delegates.LazyWithReceiver
import com.github.ericytsang.research2016.propositionallogic.Proposition.Operand
import com.github.ericytsang.research2016.propositionallogic.Proposition.Operator
import java.io.Serializable
import java.util.LinkedHashMap
import java.util.LinkedHashSet

/**
 * this class is the component in the composite oo design pattern.
 */
sealed class Proposition:Serializable
{
    companion object
    {
        /**
         * creates a [Proposition] in conjunctive normal form from the provided
         * [state], e.g. the situation {a, -b, c} would yield "a and -b and c".
         */
        fun fromState(state:State):Proposition
        {
            val propositions = state.keys.map()
            {
                basicProposition ->
                if (state[basicProposition]!!)
                {
                    basicProposition
                }
                else
                {
                    Not(basicProposition)
                }
            }
            return And.make(propositions) ?: tautology
        }

        fun makeDnf(states:Iterable<State>):Proposition
        {
            return Or.make(states.map {Proposition.fromState(it)}) ?: contradiction
        }

        /**
         * generates a [List] of [Variable]s of length [numPropositions]. each
         * generated [Variable]'s [String] is randomly chosen from
         * [basicPropositionStrings].
         */
        fun makeRandom(basicPropositionStrings:List<String>,numPropositions:Int):List<Variable>
        {
            return (1..numPropositions).map {Variable.fromString(basicPropositionStrings.getRandom())}
        }
    }

    /**
     * returns the expression as a human-readable string.
     */
    abstract override fun toString():String

    /**
     * returns a hashcode computed from the structure of the formula tree and
     * types of the nodes in the formula tree.
     */
    override fun hashCode():Int = toString().hashCode()

    /**
     * returns true when this [Proposition] structurally equals to [other].
     *
     * @param other [Proposition] to compare with for structural equality.
     */
    override fun equals(other:Any?):Boolean = other is Proposition && other.toString() == toString()

    /**
     * [List] of children of this node.
     */
    abstract val children:List<Proposition>

    /**
     * a [Proposition] that cannot be further decomposed into more [Proposition]
     * objects; a leaf node.
     *
     * @param friendly the string returned when [toString] is called.
     */
    abstract class Operand(val friendly:String):Proposition()
    {
        /**
         * returns the truth value of this [Operand] as per [state].
         */
        abstract fun truthValue(state:State):Boolean

        /**
         * returns [friendly].
         */
        override fun toString():String = friendly.toString()

        /**
         * [List] of children of this node. leaf nodes do not have children; an
         * empty [List] is returned in this implementation.
         */
        override val children:List<Proposition> = emptyList()
    }

    /**
     * a logical connective that connects its [children] together forming a
     * compound sentence; an internal node.
     */
    abstract class Operator(override val children:List<Proposition>):Proposition()
    {
        /**
         * returns the [Boolean] truth value of the operation for the given
         * [operands].
         *
         * @param operands truth value of ordered operands to operate on.
         */
        abstract fun operate(operands:List<Boolean>):Boolean

        /**
         * returns the truthiness of the operation for the given [operands]; the
         * probability that this operation would return true.
         *
         * @param operands probability of truth of ordered operands to operate
         * on.
         */
        abstract fun operate(operands:List<Double>):Double
    }

    /**
     * returns all the basic propositions in this [Proposition].
     */
    val variables:Set<Variable> by lazy()
    {
        return@lazy if (this is Variable)
        {
            setOf(this)
        }
        else
        {
            children.flatMap {it.variables}.toSet()
        }
    }

    /**
     * returns all the models of this [Proposition], i.e., all the [State] that
     * satisfy this [Proposition].
     */
    val models:Set<State> by lazy()
    {
        val branch = fun(state:State):Set<State>
        {
            val nextVariable = variables.minus(state.keys).firstOrNull()
            if (nextVariable != null)
            {
                val node1 = State.fromVariableMap(state+mapOf(nextVariable to true))
                val node2 = State.fromVariableMap(state+mapOf(nextVariable to false))
                return listOf(node1,node2)
                    .filter {truthiness(it) != 0.0}
                    .toSet()
            }
            else
            {
                return emptySet()
            }
        }

        val bounds = fun(state:State):Bounds
        {
            return Bounds(1.0,Math.floor(truthiness(state)))
        }

        // returns true if the situation is a solution; false otherwise
        val checkSolution = fun(state:State):Boolean
        {
            return truthiness(state) == 1.0 && state.keys.containsAll(variables)
        }

        val iterator = object:AbstractIterator<State>()
        {
            val unbranchedSituations = mutableMapOf(State.fromVariableMap(emptyMap()) to rootNodeMetaData)
            override fun computeNext()
            {
                val next = branchAndBound(unbranchedSituations,branch,bounds,checkSolution)
                if (next == null)
                {
                    done()
                }
                else
                {
                    setNext(next)
                }
            }
        }

        return@lazy IteratorToSetAdapter(iterator)
    }
}

infix fun Proposition.isSubsetOf(that:Proposition):Boolean
{
    return (that.not and this).isContradiction
}

/**
 * returns true when every model of this is satisfied by a model of that, and
 * every model of that satisfies a model of this.
 */
infix fun Proposition.isSatisfiedBy(that:Proposition):Boolean
{
    // this is a subset of that
    return (that isSubsetOf this) &&
        // and for each model of this there is a model of that which satisfies it
        this.models.all {(Proposition.fromState(it) and that).isSatisfiable}
}

/**
 * returns true if there is at least one model for this [Proposition]; false
 * otherwise.
 */
val Proposition.isSatisfiable:Boolean get()
{
    return models.isNotEmpty()
}

/**
 * returns true if this [Proposition] would [evaluate] to true in any situation;
 * false otherwise.
 */
val Proposition.isTautology:Boolean get()
{
    return !not.isSatisfiable
}

/**
 * returns true if this [Proposition] would [evaluate] to false in any
 * situation; false otherwise.
 */
val Proposition.isContradiction:Boolean get()
{
    return !isSatisfiable
}

/**
 * returns the truth value of this [Proposition] for the given [state].
 */
fun Proposition.evaluate(state:State):Boolean = when (this)
{
    is Operand -> truthValue(state)
    is Operator -> operate(children.map {it.evaluate(state)})
}

/**
 * returns the [truthiness] of this [Proposition] for the given [state].
 * if the proposition is more likely to be true for the given [state], it
 * will return a value closer to 1.0. if the value is more likely to be false,
 * it will return a value closer to 0.0. if the returned value is 1 or 0,
 * [evaluate] should return true or false respectively if it is passed the same
 * [state], assuming that there are no missing value mappings.
 */
fun Proposition.truthiness(state:State):Double = when (this)
{
    is Variable ->
    {
        if (this in state.keys)
        {
            if (truthValue(state)) 1.0 else 0.0
        }
        else
        {
            0.5
        }
    }
    is Operand ->
    {
        when
        {
            this == tautology -> 1.0
            this == contradiction -> 0.0
            else -> throw IllegalArgumentException("unknown atomic proposition! $this")
        }
    }
    is Operator -> operate(children.map {it.truthiness(state)})
}

/**
 * returns a logically equivalent [Proposition] in full disjunctive normal form.
 *
 * (a and b and c) or (a and -b and -c)
 */
fun Proposition.toFullDnf():Proposition
{
    // list of literals (or negations of literals) and'd together
    val literalConjunctions = models
        .mapNotNull()
        {
            situation ->
            situation.keys
                // return the proposition if it is mapped to true, and the
                // negation of it otherwise
                .map {if (situation[it]!!) it else it.not}
                // and everything together
                .let {if (it.isNotEmpty()) And.make(it) else tautology}
        }
    return Or.make(literalConjunctions) ?: contradiction
}

/**
 * returns a logically equivalent [Proposition] in disjunctive normal form...
 * simplified as much as possible as if one was using Karnaugh maps...
 *
 * (a and b) or -a
 */
fun Proposition.toDnf():Proposition
{
    val hammingDistance = fun(state1:State,state2:State):Int
    {
        return state1.keys.count {state1[it] != state2[it]}
    }

    val basicPropositionsToSituations = LinkedHashMap(models.groupBy {it.keys}.mapValues {it.value.toMutableSet()})
    val unprocessedKeys = basicPropositionsToSituations.keys.toMutableSet()

    while (unprocessedKeys.isNotEmpty())
    {
        val basicPropositions = unprocessedKeys.maxBy {it.size}!!
        val situations = basicPropositionsToSituations[basicPropositions]!!
        var state1:State? = null
        var state2:State? = null
        loop@for (situationA in situations)
        {
            for (situationB in situations)
            {
                if (hammingDistance(situationA,situationB) == 1)
                {
                    state1 = situationA
                    state2 = situationB
                    break@loop
                }
            }
        }
        if (state1 == null && state2 == null)
        {
            unprocessedKeys.remove(basicPropositions)
        }
        else
        {
            situations.remove(state1)
            situations.remove(state2)

            val commonMappings = state1!!.entries.filter {state1!![it.key] == state2!![it.key]}
            val newSituation = State.fromVariableMap(commonMappings.associate {it.key to it.value})
            basicPropositionsToSituations.getOrPut(newSituation.keys,{mutableSetOf()}).add(newSituation)
            unprocessedKeys.add(newSituation.keys)
        }
    }

    // list of literals (or negations of literals) and'd together
    val literalConjunctions = basicPropositionsToSituations
        .flatMap {it.value}
        .mapNotNull()
        {
            situation ->
            situation.keys
                // return the proposition if it is mapped to true, and the
                // negation of it otherwise
                .map {if (situation[it]!!) it else it.not}
                // and everything together
                .let {if (it.isNotEmpty()) And.make(it) else tautology}
        }
    return Or.make(literalConjunctions) ?: contradiction
}
